pkg/controllers/securitypolicy_associated_controller.go

```diff
@@ -0,0 +1,241 @@
+}
+
+func (e *EnqueueRequestForPod) Update(updateEvent event.UpdateEvent, l workqueue.RateLimitingInterface) {
+	e.Raw(updateEvent, l)
+}
+
+func (e *EnqueueRequestForPod) Delete(deleteEvent event.DeleteEvent, l workqueue.RateLimitingInterface) {
+	e.Raw(deleteEvent, l)
+}
+
+func (e *EnqueueRequestForPod) Generic(genericEvent event.GenericEvent, l workqueue.RateLimitingInterface) {
+	e.Raw(genericEvent, l)
+}
+
+func (e *EnqueueRequestForPod) Raw(evt interface{}, q workqueue.RateLimitingInterface) {
+	var pods []v1.Pod
+	var obj client.Object
+	var oldObj client.Object
+
+	switch et := evt.(type) {
+	case event.CreateEvent:
+		obj = et.Object.(*v1.Pod)
+	case event.UpdateEvent:
+		obj = et.ObjectNew.(*v1.Pod)
+		oldObj = et.ObjectOld.(*v1.Pod)
+		pods = append(pods, *oldObj.(*v1.Pod))
+	case event.DeleteEvent:
+		obj = et.Object.(*v1.Pod)
+	case event.GenericEvent:
+		obj = et.Object.(*v1.Pod)
+	default:
+		log.Error(nil, "Unknown event type", "event", evt)
+	}
+	pod := obj.(*v1.Pod)
+	pods = append(pods, *pod)
+	err := reconcileSecurityPolicy(e.Client, pods, q)
+	if err != nil {
+		log.Error(err, "Failed to reconcile security policy")
+	}
+}
+
+func getAllPodPortNames(pods []v1.Pod) sets.String {
+	podPortNames := sets.NewString()
+	for _, pod := range pods {
+		for _, container := range pod.Spec.Containers {
+			for _, port := range container.Ports {
+				if port.Name != "" {
+					podPortNames.Insert(port.Name)
+				}
+			}
+		}
+	}
+	return podPortNames
+}
+
+func reconcileSecurityPolicy(client client.Client, pods []v1.Pod, q workqueue.RateLimitingInterface) error {
+	podPortNames := getAllPodPortNames(pods)
+	log.V(1).Info("pod named port", "podPortNames", podPortNames)
+	spList := &v1alpha1.SecurityPolicyList{}
+	err := client.List(context.Background(), spList)
+	if err != nil {
+		log.Error(err, "failed to list security policy")
+		return err
+	}
+
+	for _, securityPolicy := range spList.Items {
+		shouldReconcile := false
+		for _, rule := range securityPolicy.Spec.Rules {
+			for _, port := range rule.Ports {
+				if port.Port.Type == intstr.String {
+					if podPortNames.Has(port.Port.StrVal) {
+						shouldReconcile = true
+					}
+				}
+			}
+		}
+		if shouldReconcile {
+			log.Info("reconcile it", "namespace", securityPolicy.Namespace, "name", securityPolicy.Name)
+			q.Add(reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      securityPolicy.Name,
+					Namespace: securityPolicy.Namespace,
+				},
+			})
+		}
+	}
+	return nil
+}
+
+var PredicateFuncsPod = predicate.Funcs{
+	CreateFunc: func(e event.CreateEvent) bool {
+		if p, ok := e.Object.(*v1.Pod); ok {
+			return checkPodHasNamedPort(*p, "create")
+		}
+		return false
+	},
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		if p, ok := e.ObjectOld.(*v1.Pod); ok {
+			if checkPodHasNamedPort(*p, "update") {
+				return true
+			}
+		}
+		if p, ok := e.ObjectNew.(*v1.Pod); ok {
+			if checkPodHasNamedPort(*p, "update") {
+				return true
+			}
+		}
```

Comment:
What's the meanings of ObjectOld and ObjectNew? Could you please add a comment to explain why we need to do the check L180~L191? For example, in which case the return value will be false?

Reply:
ObjectOld and ObjectNew are defined by controller-runtime. Others I will comment on.